src/QueryPipeline/QueryPipelineBuilder.cpp

```diff
@@ -389,9 +389,11 @@ std::unique_ptr<QueryPipelineBuilder> QueryPipelineBuilder::joinPipelines(
     auto lit = left->pipe.output_ports.begin();
     auto rit = right->pipe.output_ports.begin();
 
+    auto output_header = JoiningTransform::transformHeader(left->getHeader(), join);
```

Comment:
Do I understand correctly that this output header is already calculated/updated at JoinStep level?
https://github.com/ClickHouse/ClickHouse/blob/fdd2db4aceac392e92665aefc4c74180d2bf9ad0/src/Processors/QueryPlan/JoinStep.cpp#L24-L27
https://github.com/ClickHouse/ClickHouse/blob/fdd2db4aceac392e92665aefc4c74180d2bf9ad0/src/Processors/QueryPlan/JoinStep.cpp#L46-L48
If so, then it looks like it could be provided from JoinStep

Reply:
Here we join two pipelines directly. So we don't have `JoinStep` available.

Reply:
Thanks for clarification

Reply:
@amosbird but we do call `QueryPipelineBuilder::joinPipelines` only from `JoinStep` isn't it?

https://github.com/ClickHouse/ClickHouse/blob/fdd2db4aceac392e92665aefc4c74180d2bf9ad0/src/Processors/QueryPlan/JoinStep.cpp#L30-L36